grr/server/grr_response_server/flows/general/osquery_test.py

```diff
@@ -147,48 +152,108 @@ def testFailure(self):
       self._RunQuery("UPDATE time SET day = -1;")
 
 
-class FakeOsqueryFlowTest(flow_test_lib.FlowTestsBaseclass):
+def _CreateDictTable(column_number, row_number) -> List[OrderedDict]:
+  def CellValue(col_number, row_number):
+    return (f"col-{col_number}", f"col-{col_number}, row-{row_number}")
+
+  return [
+    OrderedDict(
+      CellValue(col_number, row_number) for col_number in range(column_number)
+    ) for row_number in range(row_number)
+  ]
```

Comment:
This makes the test code hard to follow—avoid using any helpers in the test code if possible. Here, you don't care about the values at all (whether they are unique or not) and the number of columns is not too high, so you can prepare a readable test data yourself, e.g. `[{ "foo": "quux", "baz": "thud", "bar": "norf" }] * 42`.

Reply:
Another thing is that the `[{ "foo": "quux", "baz": "thud", "bar": "norf" }] * 42` sample input wouldn't capture errors such as reordering of rows, as all of them are the same.

Reply:
> However, what if a test fails because some cell values disagree?

If values are relevant they should not be introduced next to the place where you make assertions, otherwise you make assumption on what the format the helper is using, forcing the reader to familiarize themselves with the helper code (which is bad for the reasons explained in the next paragraph).

> Isn't it readable enough if the complexity is hidden behind helper functions?

If the reader has to jump around the code in order to understand what it is happening, it is not readable. Tests should read like linear scenario and be obvious. Here, I had to spend significant amount of time to understand what this helper is doing.

Moreover, it is not only about the readability. The more complex the helper is, the more likely you are to end up in a scenario where your generated data is invalid, e.g. it is empty and your tests passes trivially. Ultimately, in such cases, you should write tests for the helper itself (and at this point you not only introduce complexity, but also you have written more code than just inlining some input in the test code itself).

> Another thing is that the [{ "foo": "quux", "baz": "thud", "bar": "norf" }] * 42 sample input wouldn't capture errors such as reordering of rows, as all of them are the same.

Tests should be minimal and written for a particular purpose. If you want to test whether your code preserves the order, you should write small test case with pre-made order-sensitve input. If you want to test whether the code works with large input, then you can do `[single_row] * 42` and it still tests what it should.